### PR TITLE
[fuzz] Add a concept of a visitor to gen. validated input.

### PR DESCRIPTION
**Pull request by vehre-x41**
Commit Message: [fuzz] Add a concept of a visitor to gen. validated input.
Additional Description:
During fuzzing the generated input often violates the validation constraints already in the proto-files.
This visitor tries to use these constraints as best as possible to massage the input in a way that it passes
input validation. 

Risk Level: low
Testing: n/a
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Comment by adisuissa** [comment]
Thanks!
Before going into the details, are there indications how much this can improve fuzzer execution?

**Review by adisuissa — COMMENTED** [review]
Thanks.
I think it will help guiding the fuzzer, I'm just concerned about subtle bugs that this code may introduce that will be more challenging to debug, and would not like to have tests for tests.

I've got a high-level question: what will happen if the `PostProcessorRegistration` executes the input validation, and if it fails, changes the input to a known valid config. Assuming the hashing of the input happens after, I'm assuming the fuzzing infrastructure will short-circuit and try another input. If so, will this be sufficient throughput improvement at the expense of guiding the input in a more granular way?

File: source/common/protobuf/visitor.cc
Comment (vehre-x41) [reply source/common/protobuf/visitor.cc:149]: Yes, of course can I move it there. The initial plan was to just adapt the `traverseMessageWorker(..)`, but that unfortunately is not feasible.

Just to get you right: `protobuf_utility.h` does not exist yet. You want me to create that file, or do you mean to add to `test/fuzz/utility.h|cpp`?

Comment (adisuissa) [reply source/common/protobuf/visitor.cc:149]: IMHO a new file is better, as it may not be needed for general fuzzers, and avoiding linking the protobuf-traversal code will reduce the binary size (a bit).

Comment (vehre-x41) [reply source/common/protobuf/visitor.cc]: I have refactored the visitor now. The old one is restored, but because both needed some support routines, I also factored them into a visitor_helper.h|cc in source/common/protobuf. The new visitor's implementation is in test/fuzz/mutable_visitor.h|cc.

Is that what you had in mind? Or does it go at least into the right direction?

File: test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc
Comment (vehre-x41) [reply test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc]: I have removed them. When they are desired again, one can always add them back from the history.

**Comment by vehre-x41** [comment]
Mmhh, looks like CI is not running. Is it because the branch's base is so old? Rebasing does not work either, because that fails push-checks in files I haven't touched... What to do?

**Comment by yanavlasov** [comment]
It looks like CI is blocked because of existing conflicts. Please merge main and resolve conflicts to unblock CI.

**Comment by vehre-x41** [comment]
Ok, I merged in main but had to fix code style in three files or the pre-push hook would block. The fixes are accumulated in the https://github.com/envoyproxy/envoy/pull/26141/commits/1bbc58df9e2379ce735e802c83de7aabb1a685e9 commit. The fixes were done automatically either by the python script itself or as proposed using a patch generated from the proto_sync tool.

**Comment by vehre-x41** [comment]
This gets ridiculous. My local pre-push hook wants the three files changed in the one direction, and CI wants to have them changed back. Rebased now to get this clean once and for all.

**Comment by vehre-x41** [comment]
The CI is now finally happy. Any comments?

**Comment by vehre-x41** [comment]
/retest

due to CI internal issue

File: test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc
Comment (adisuissa) [inline test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc:36]: Is this to allow faster creation of valid configs? Can you add a comment why this is done here (and give some insight why in network_writefilter_fuzz_test it is not).

**Review by adisuissa — COMMENTED** [review]
Thanks!
Minor question, but otherwise LGTM

File: test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc
Comment (vehre-x41) [reply test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc]: Clearing the config is to allow unpacking of `value` correctly. Imagine that a config was present for an RBAC filter and now some other filter is chosen. Unpacking that config might result into some "funny" values, if it works at all.

I am doing this conditionally now, i.e., only when the filter's name has changed.

Applied the changes to both network_read-/writerfilter_fuzz_tests.

File: test/fuzz/validated_input_generator.cc
Comment (adisuissa) [inline test/fuzz/validated_input_generator.cc:17]: It may be better to remove this macro and set it to info - less global clobber.
If needed, the test log can probably be searched for "pgv::".

**Review by adisuissa — COMMENTED** [review]
Thanks! One small nit, and that's good to go

File: test/fuzz/validated_input_generator.cc
Comment (vehre-x41) [reply test/fuzz/validated_input_generator.cc]: Fixed

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!
Needs a senior maintainer LGTM on the core-code refactor.
/assign-from @envoyproxy/senior-maintainers

**Comment by vehre-x41** [comment]
/retest

due to flaky mobile_core step

**Comment by adisuissa** [comment]
@ggreenway can you PTAL at the (rather small) refactor in the files in source/common/protobuf ?

**Comment by adisuissa** [comment]
@vehre-x41 the PR needs to merge main.

**Review by ggreenway — DISMISSED** [review]
Core code refactor LGTM

**Comment by ggreenway** [comment]
merge conflict

/wait

**Review by adisuissa — APPROVED** [review]
LGTM!
Thank you @vehre-x41 for working on this!
Thanks @ggreenway for the review and approval of core-code changes.

**Comment by vehre-x41** [comment]
Unfortunately yes, I'll look into that. Curiously on first glance, a filter keeps repeating in the corpus.

**Comment by adisuissa** [comment]
> Irritatingly that corpus run stand-alone passes. Yes, the config in there is long, but shouldn't be a reason to abort, right?

It may be that the execution env is a bit different (different frame cap in oss-fuzz compared to the local setting).
An alternative will be to replace the invalid sub-message with a predefined valid one. So instead of traversing the structure, validate if it's valid, and if not, replace it. This can be done after 2-3 levels of going down the rabbit-hole.

**Comment by adisuissa** [comment]
> An easy alternative would be to limit the traversal depth from currently 40 to for example 20. What do you think?

Sure.

> We should also think about, if limiting the depth makes sense at all. When the configuration is an attack surface and envoy does not limit the depth itself, then an attacker could abuse it to crash envoy. Any opinion on that?

I'm less worried about the depth, because:
1. The main issues we are looking for are closer to things that do not work (e.g., crash), and we are not focused on very bad configuration. In other words, is there a valid configuration that can lead to subtle bugs in runtime.
2. Similar to depth, there's also a problem of large config size, but this is not part of what we are trying to detect.

**Pull request merged**
